File: java/lib/src/main/java/dev/argon/esexpr/codecs/StringCodec.java
package dev.argon.esexpr.codecs;

import dev.argon.esexpr.*;

import java.util.Set;

/**
 * A codec for string values.
 */
public class StringCodec implements ESExprCodec<String> {
	private StringCodec() {}

	/**
	 * A codec for string values.
	 */
	
	@ESExprCodecTags(scalar = { ESExprTag.Scalar.STR })
	public static final ESExprCodec<String> INSTANCE = new StringCodec();

	@Override
	public ESExprTagSet tags() {
		return ESExprTagSet.of(ESExprTag.STR);
	}

	@Override
	public boolean isEncodedEqual(String x, String y) {
		return x.equals(y);
	}

	@Override
	public ESExpr encode(String value) {
		return new ESExpr.Str(value);
	}

	@Override
	public String decode(ESExpr expr, FailurePath path) throws DecodeException {
		if(expr instanceof ESExpr.Str(var s)) {
			return s;
		}
		else {
			throw new DecodeException("Expected a string value", path);
		}
	}
}

File: dotnet/ESExpr.Runtime/Codecs/StringCodec.cs
using System.Collections.Generic;

namespace ESExpr.Runtime.Codecs;

public sealed class StringCodec : IESExprCodec<string> {
	public ESExprTagSet Tags => ESExprTagSet.Create([new ESExprTag.Str()]);

	public bool IsEncodedEqual(string a, string b) => a == b;

	public Expr Encode(string value) => new Expr.Str(value);

	public string Decode(Expr expr, DecodeFailurePath path) {
		if(expr is Expr.Str(var s)) {
			return s;
		}
		else {
			throw new DecodeException("Expected a string value", path);
		}
	}
}
